php/class-customize-snapshot-manager.php

```diff
@@ -0,0 +1,403 @@
+		}
+	}
+
+	/**
+	 * Create the custom post type.
+	 *
+	 * @access public
+	 */
+	public function create_post_type() {
+		$args = array(
+			'labels' => array(
+				'name' => __( 'Customize Snapshots', 'customize-widgets-plus' ),
+				'singular_name' => __( 'Customize Snapshot', 'customize-widgets-plus' ),
+			),
+			'public' => false,
+			'capability_type' => 'post',
+			'map_meta_cap' => true,
+			'hierarchical' => false,
+			'rewrite' => false,
+			'delete_with_user' => false,
+			'supports' => array( 'title', 'author', 'revisions' ),
+		);
+
+		register_post_type( self::POST_TYPE, $args );
+	}
+
+	/**
+	 * Enqueue styles & scripts for the Customizer.
+	 *
+	 * @action customize_controls_enqueue_scripts
+	 */
+	public function enqueue_scripts() {
+		$handle = 'customize-snapshot';
+
+		// Enqueue styles.
+		wp_enqueue_style( $this->plugin->style_handles[ $handle ] );
+
+		// Enqueue scripts.
+		wp_enqueue_script( $this->plugin->script_handles[ $handle ] );
+
+		// Script data array.
+		$exports = array(
+			'nonce' => wp_create_nonce( self::AJAX_ACTION ),
+			'action' => self::AJAX_ACTION,
+			'uuid' => $this->snapshot->uuid(),
+			'is_preview' => $this->snapshot->is_preview(),
+			'scope' => ( isset( $_GET['scope'] ) ? $_GET['scope'] : 'dirty' ),
+			'i18n' => array(
+				'saveButton' => __( 'Save', 'customize-widgets-plus' ),
+				'cancelButton' => __( 'Cancel', 'customize-widgets-plus' ),
+				'shareButton' => __( 'Share URL to preview', 'customize-widgets-plus' ),
+				'updateMsg' => __( 'Clicking "Save" will update the current snapshot.', 'customize-widgets-plus' ),
+				'errorMsg' => __( 'The snapshot could not be saved.', 'customize-widgets-plus' ),
+				'previewTitle' => __( 'Preview URL', 'customize-widgets-plus' ),
+				'formTitle' => ( $this->snapshot->is_preview() ?
+					__( 'Update Snapshot', 'customize-widgets-plus' ) :
+					__( 'Snapshot Scope', 'customize-widgets-plus' )
+				),
+				'dirtyLabel' => __( 'Diff Snapshot (preview dirty settings)', 'customize-widgets-plus' ),
+				'fullLabel' => __( 'Full Snapshot (preview all settings)', 'customize-widgets-plus' ),
+			),
+		);
+
+		// Export data to JS.
+		wp_scripts()->add_data(
+			$this->plugin->script_handles[ $handle ],
+			'data',
+			sprintf( 'var _customizeWidgetsPlusCustomizeSnapshot = %s;', wp_json_encode( $exports ) )
+		);
+	}
+
+	/**
+	 * Get the Customize_Snapshot instance.
+	 *
+	 * @return Customize_Snapshot
+	 */
+	public function snapshot() {
+		return $this->snapshot;
+	}
+
+	/**
+	 * Save a snapshot.
+	 *
+	 * @param WP_Customize_Manager $manager WP_Customize_Manager instance.
+	 * @param string $status The post status.
+	 * @return null|WP_Error
+	 */
+	public function save( \WP_Customize_Manager $manager, $status = 'draft' ) {
+		$new_setting_ids = array_diff( array_keys( $this->post_data ), array_keys( $manager->settings() ) );
+		$manager->add_dynamic_settings( wp_array_slice_assoc( $this->post_data, $new_setting_ids ) );
```

Comment:
@westonruter Calling `add_dynamic_settings` here throws errors if you've added widgets or menus to the snapshot. I suppose it will throw errors for any dynamic settings. I'm not sure what's happening because the setting IDs are being passed correctly. I hacked Core to see what was being passed and the data looked fine while I was debugging but when you let it continue it throws an illegal index or null error.